{% extends "layout.html" %}

{% block in_head %}
<script src="https://cdn.jsdelivr.net/npm/d3@^5.16.0" type="text/javascript"></script>
<script src="https://unpkg.com/d3-tip@0.9.1"></script>
<script src="{{ url_for('.static', filename='vendor/stream_table-1.1.1.min.js') }}" type="text/javascript"></script>
<script src="{{ url_for('.static', filename='pheno.js') }}" type="text/javascript"></script>
{% if show_correlations %}
  {#  TODO: Better dependency management #}
  <link href="https://unpkg.com/tabulator-tables@4.2.2/dist/css/bootstrap/tabulator_bootstrap.min.css" rel="stylesheet">
  <script type="text/javascript" src="https://unpkg.com/tabulator-tables@4.2.2/dist/js/tabulator.min.js"></script>
{% endif %}

<script type="text/javascript">
  window.model.tooltip_underscoretemplate = {{ tooltip_underscoretemplate|tojson }};
  window.phenocode = {{ phenocode|tojson|safe }};
  window.phenocode_male = {{ phenocode|tojson|safe }} + ".male";
  window.phenocode_female = {{ phenocode|tojson|safe }} + ".female";
  window.pheno_male = {{ pheno_male | tojson }};
  window.pheno_female = {{ pheno_female | tojson }};
  window.model.sex_stratified = {{ sex_stratified | tojson }};
  window.model.show_correlations = {{ show_correlations | tojson }};
  window.model.correlations_url = {{ url_for('.api_pheno_correlations', phenocode=pheno['phenocode'] ) | tojson }};
  window.model.pheno_correlations_pvalue_threshold = {{ pheno_correlations_pvalue_threshold | tojson }};
  
  $.getJSON(window.model.urlprefix + "/api/manhattan/pheno/" + window.phenocode + ".json")
  .done(function(data) {
      window.debug.manhattan = data;
      create_gwas_plot(data.variant_bins, data.unbinned_variants);
      populate_streamtable(data.unbinned_variants);
  })
  .fail(function() {
    console.log("Manhattan XHR failed");
  });

  if (window.model.sex_stratified && window.pheno_male != null && window.pheno_female != null){

    $.when(
      $.getJSON(window.model.urlprefix + "/api/miami/pheno/" + window.phenocode_male + ".json"),
      $.getJSON(window.model.urlprefix + "/api/miami/pheno/" + window.phenocode_female + ".json")
    ).done(function(data_male, data_female) {
        window.debug.miami = {
          male: data_male[0],
          female: data_female[0]
        };
        create_miami_plot({
          male_variant_bins : data_male[0].variant_bins, male_unbinned_variants : data_male[0].unbinned_variants,
          female_variant_bins :data_female[0].variant_bins, female_unbinned_variants : data_female[0].unbinned_variants
        });
        // we won't worry about this for now.
        // populate_streamtable(data_male.unbinned_variants);
    })
    .fail(function() {
      console.log("Miami (with male and female) XHR failed");
    });
  }

  else if (window.model.sex_stratified && window.pheno_female == null && window.pheno_male != null){
    $.when(
      $.getJSON(window.model.urlprefix + "/api/miami/pheno/" + window.phenocode_male + ".json"),
    ).done(function(data_male) {
        window.debug.miami = {
          male: data_male,
          female: null
        };
        create_miami_plot({male_variant_bins : data_male.variant_bins, male_unbinned_variants : data_male.unbinned_variants});
        // we won't worry about this for now.
        // populate_streamtable(data_male.unbinned_variants);
    })
    .fail(function() {
      console.log("Miami (with male only) XHR failed");
    });
  }
  else if (window.model.sex_stratified && window.pheno_male == null && window.pheno_female != null){
    $.when(
      $.getJSON(window.model.urlprefix + "/api/miami/pheno/" + window.phenocode_female + ".json"),
    ).done(function(data_female) {
        window.debug.miami = {
          female: data_female,
          male: null
        };
        create_miami_plot({female_variant_bins: data_female.variant_bins, female_unbinned_variants: data_female.unbinned_variants});
        // we won't worry about this for now.
        // populate_streamtable(data_male.unbinned_variants);
    })
    .fail(function() {
      console.log("Miami (with female only) XHR failed");
    });
  }

  if (window.model.sex_stratified && window.pheno_male != null && window.pheno_female != null){
    $.when(
      $.getJSON(window.model.urlprefix + "/api/qq/pheno/" + window.phenocode + ".json"),
      $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.phenocode_male + ".json"),
      $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.phenocode_female + ".json")
    ).done(function(data1, data2, data3) {
      qq_combined = data1[0]
      qq_male = data2[0]
      qq_female = data3[0]
      _.sortBy(_.pairs(qq_combined.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
        var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
        if (i === 0) { text = '<b>' + text + '</b>'; }
        text = '<br>' + text;
        $('.gc-control_both').append(text);
      });
      _.sortBy(_.pairs(qq_male.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
        var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
        if (i === 0) { text = '<b>' + text + '</b>'; }
        text = '<br>' + text;
        $('.gc-control_male').append(text);
      });
      _.sortBy(_.pairs(qq_female.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
        var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
        if (i === 0) { text = '<b>' + text + '</b>'; }
        text = '<br>' + text;
        $('.gc-control_female').append(text);
      });

      //return the max height and width to use based off the max of each .json
      var dimensions = calculate_qq_dimension(qq_combined, male_data = qq_male, female_data = qq_female);

      if (qq_combined.by_maf){
        create_qq_plot(qq_combined.by_maf, qq_combined.ci, dimensions[0], dimensions[1]);
        create_qq_plot(qq_male.by_maf, qq_male.ci, dimensions[0], dimensions[1], "male");
        create_qq_plot(qq_female.by_maf, qq_female.ci, dimensions[0], dimensions[1], "female");
      }
      else {
        create_qq_plot([{maf_range:[0,0.5], qq:qq_combined.overall.qq, count:qq_combined.overall.count}], qq_combined.ci, dimensions[0], dimensions[1]);
        create_qq_plot([{maf_range:[0,0.5], qq:qq_male.overall.qq, count:qq_male.overall.count}], qq_male.ci, dimensions[0], dimensions[1], "male");
        create_qq_plot([{maf_range:[0,0.5], qq:qq_female.overall.qq, count:qq_female.overall.count}], qq_female.ci, dimensions[0], dimensions[1], "female");
      }
      })
      .fail(function() {
        console.log("QQ (with male and female) XHR failed");
      });
    }
    else if (window.model.sex_stratified && window.pheno_female == null && window.pheno_male != null){
      $.when(
        $.getJSON(window.model.urlprefix + "/api/qq/pheno/" + window.phenocode + ".json"),
        $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.phenocode_male + ".json"),
      ).done(function(data1, data2) {
        qq_combined = data1[0]
        qq_male = data2[0]
        _.sortBy(_.pairs(qq_combined.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
          var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
          if (i === 0) { text = '<b>' + text + '</b>'; }
          text = '<br>' + text;
          $('.gc-control_both').append(text);
        });
        _.sortBy(_.pairs(qq_male.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
          var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
          if (i === 0) { text = '<b>' + text + '</b>'; }
          text = '<br>' + text;
          $('.gc-control_male').append(text);
        });

        //return the max height and width to use based off the max of each .json
        var dimensions = calculate_qq_dimension(qq_combined, male_data = qq_male);

        if (qq_combined.by_maf){
          create_qq_plot(qq_combined.by_maf, qq_combined.ci, dimensions[0], dimensions[1]);
          create_qq_plot(qq_male.by_maf, qq_male.ci, dimensions[0], dimensions[1], "male");
        }
        else {
          create_qq_plot([{maf_range:[0,0.5], qq:qq_combined.overall.qq, count:qq_combined.overall.count}], qq_combined.ci, dimensions[0], dimensions[1]);
          create_qq_plot([{maf_range:[0,0.5], qq:qq_male.overall.qq, count:qq_male.overall.count}], qq_male.ci, dimensions[0], dimensions[1], "male");
        }
        })
        .fail(function() {
          console.log("QQ (with male) XHR failed");
        });
    }
    else if (window.model.sex_stratified && window.pheno_male == null && window.pheno_female != null){
      $.when(
        $.getJSON(window.model.urlprefix + "/api/qq/pheno/" + window.phenocode + ".json"),
        $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.phenocode_female + ".json"),
      ).done(function(data1, data2) {
        qq_combined = data1[0]
        qq_female = data2[0]
        _.sortBy(_.pairs(qq_combined.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
          var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
          if (i === 0) { text = '<b>' + text + '</b>'; }
          text = '<br>' + text;
          $('.gc-control_both').append(text);
        });
        _.sortBy(_.pairs(qq_female.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
          var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
          if (i === 0) { text = '<b>' + text + '</b>'; }
          text = '<br>' + text;
          $('.gc-control_female').append(text);
        });

        //return the max height and width to use based off the max of each .json
        var dimensions = calculate_qq_dimension(qq_combined, female_data = qq_female);

        if (qq_combined.by_maf){
          create_qq_plot(qq_combined.by_maf, qq_combined.ci, dimensions[0], dimensions[1]);
          create_qq_plot(qq_female.by_maf, qq_female.ci, dimensions[0], dimensions[1], "female");
        }
        else {
          create_qq_plot([{maf_range:[0,0.5], qq:qq_combined.overall.qq, count:qq_combined.overall.count}], qq_combined.ci, dimensions[0], dimensions[1]);
          create_qq_plot([{maf_range:[0,0.5], qq:qq_female.overall.qq, count:qq_female.overall.count}], qq_female.ci, dimensions[0], dimensions[1], "female");
        }
        })
        .fail(function() {
          console.log("QQ (with female) XHR failed");
        });
    } else {
      $.getJSON(window.model.urlprefix + "/api/qq/pheno/" + window.phenocode + ".json")
      .done(function(data) {
      window.debug.qq = data;
      _.sortBy(_.pairs(data.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
        var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
        if (i === 0) { text = '<b>' + text + '</b>'; }
        text = '<br>' + text;
        $('.gc-control_both').append(text);
      });
      if (data.by_maf)
        create_qq_plot(data.by_maf, data.ci, null, null);
      else
        create_qq_plot([{maf_range:[0,0.5], qq:data.overall.qq, count:data.overall.count}], data.ci, dimension[0], dimension[1]);
      })
      .fail(function() {
        console.log("QQ XHR failed");
      });
    }

</script>
<style type="text/css">
.d3-tip {
  line-height: 1.4;
  padding: 12px;
  background: rgba(0, 0, 0, 0.8);
  color: #fff;
  border-radius: 4px;
  pointer-events: none;
}
/* Creates a small triangle extender for the tooltip */
.d3-tip:after {
  display: inline;
  font-size: 10px;
  width: 100%;
  line-height: 1;
  color: rgba(0, 0, 0, 0.8);
  position: absolute;
  pointer-events: none;
}
/* Northward tooltips */
.d3-tip.n:after {
  content: "\25BC";
  margin: -3px 0 0 0;
  top: 100%;
  left: 0;
  text-align: center;
}

#manhattan_plot_container {
  min-width: 700px;
}
#manhattan_plot_container .axis > path.domain {
  stroke-width: 2px;
  stroke: #666;
  fill: none;
}
#manhattan_plot_container .axis g.tick line {
  stroke: #666;
}

#miami_plot_container {
  min-width: 700px;
}
#miami_plot_container .axis > path.domain {
  stroke-width: 2px;
  stroke: #666;
  fill: none;
}
#miami_plot_container .axis g.tick line {
  stroke: #666;
}

#qq_plot_container_both .axis path.domain {
  stroke: black;
  fill: none;
  text-align: center;
}
#qq_plot_container_both g.tick line {
  stroke: #666;
  opacity: 0.3;
}

#qq_plot_container_male .axis path.domain {
  stroke: black;
  fill: none;
  text-align: center;
}
#qq_plot_container_male g.tick line {
  stroke: #666;
  opacity: 0.3;
}

#qq_plot_container_female .axis path.domain {
  stroke: black;
  fill: none;
  text-align: center;
}
#qq_plot_container_female g.tick line {
  stroke: #666;
  opacity: 0.3;
}

.pheno-info > p {
  margin-bottom: 0;
}

.section-link {
  margin-right: 20px;
  display: inline;
}

.tabulator .tabulator-footer .tabulator-page.active {
  background-color: #337ab7;
  color: white;
}

.parent {
  background: mediumpurple;
  padding: 1rem
}

.child {
  border: 1px solid indigo;
  padding: 1rem
}

.inline-block-child {
  display: inline-block;
}

</style>
{% endblock %}


{% block contained %}
{% for key in ['num_cases', 'num_controls', 'num_samples'] %}
 {% if key in pheno %}
  {% if '<' in pheno[key]|string %}
    <div class="alert alert-danger" role="alert"><b>Warning:</b> This phenotype has {{ pheno[key] }} {{ key.replace('num_','') }}.</div>
  {% elif 0 < pheno[key]|int < 200 %}
    <div class="alert alert-danger" role="alert"><b>Warning:</b> This phenotype only has {{ pheno[key] }} {{ key.replace('num_', '') }}.</div>
  {% endif %}
 {% endif %}
{% endfor %}
<br>
<div class="row">
    <div class="col-xs-12">
        <h1 style="margin-top:0">{% include "pheno/h1.html" %}</h1>
    </div>
</div>
<div class="row">
    <div class="pheno-info col-xs-12 col-sm-8">
        {% if 'num_cases' in pheno and pheno['num_cases'] != "" %}
          <p><b>{{ pheno.num_cases }}</b> cases, <b>{{ pheno.num_controls }}</b> controls</p>
        {% elif 'num_samples' in pheno %}
          <p><b>{{ pheno.num_samples }}</b> samples</p>
        {% endif %}
        {% if pheno.category %}
          <p>Category: <b>{{ pheno.category}}</b></p>
        {% endif %}
        {% include "pheno/info.html" %}
    </div>
    <div class="pheno-info col-xs-12 col-sm-4">
      {% if config.DOWNLOAD_PHENO_SUMSTATS_BUTTON %}
      <span class="pull-right">
        <a class="btn btn-default" href="{{ url_for('.download_pheno', phenocode=phenocode) }}">Download summary statistics</a>
        {% if sex_stratified %}
          {% if pheno_male != None %}<a class="btn btn-default" href="{{ url_for('.download_pheno_sex_stratified', phenocode=phenocode, sex='male') }}">Download male summary statistics</a>
          {% endif %}
          {% if pheno_female != None %}<a class="btn btn-default" href="{{ url_for('.download_pheno_sex_stratified', phenocode=phenocode, sex='female') }}">Download female summary statistics</a>
          {% endif %}
        {% endif %}
      </span>
        
      {% endif %}
      {% if show_manhattan_filter_button %}
      <span class="pull-right" style="margin-left:1em;margin-right:1em">
        <a class="btn btn-default" href="{{ url_for('.pheno_filter_page', phenocode=phenocode) }}">Filter Variants</a>
      </span>
      {% endif %}
    </div>
</div>
<div class="row">
  <div class="col-xs-12">
    <ul>
      <li class="section-link"><a href="#manhattan">Manhattan</a></li>
      {% if sex_stratified %}<li class="section-link"><a href="#miami">Miami</a></li>{% endif %}
      <li class="section-link"><a href="#qq">QQ</a></li>
      {% if show_correlations %}<li class="section-link"><a href="#correlations">Correlated phenotypes</a></li>{% endif %}
    </ul>
  </div>
</div>
<div class="row">
    <div class="col-xs-12" id="manhattan">
        <div id="manhattan_plot_container">
        </div>
    </div>
</div>

{% if sex_stratified %}
  <div class="pheno-info col-xs-12 col-sm-8">
    {% if pheno_male != None %}
      {% if 'num_cases' in pheno_male and pheno_male['num_cases'] != "" %}
      <p><b>{{ pheno_male.num_cases }}</b> male cases, <b>{{ pheno_male.num_controls }}</b> male controls</p>
      {% elif 'num_samples' in pheno_male %}
      <p><b>{{ pheno_male.num_samples }}</b> male samples</p>
      {% endif %}
    {% endif %}
    {% if pheno_female != None %}
      {% if 'num_cases' in pheno_female and pheno_female['num_cases'] != "" %}
        <p><b>{{ pheno_female.num_cases }}</b> female cases, <b>{{ pheno_female.num_controls }}</b> female controls</p>
      {% elif 'num_samples' in pheno_female %}
        <p><b>{{ pheno_female.num_samples }}</b> female samples</p>
      {% endif %}
    {% endif %}
    {% if pheno.category %}
      <p>Category: <b>{{ pheno.category}}</b></p>
    {% endif %}
    {% include "pheno/info.html" %}
  </div>
{% endif %}

{% if sex_stratified %}
<div class="row" >
    <div class="col-xs-12" id="miami" >
        <div id="miami_plot_container">
        </div>
    </div>
</div>
{% endif %}

<div class="row">
  <div class="col-xs-12 col-md-8">
    <h3>Top Loci:</h3>
  </div>
</div>
<div class="row">
  <div class="col-xs-12 col-sm-8 col-md-6">
    <input type="text" class="form-control" id="search" placeholder="Search... &quot;TCF7L2&quot;, &quot;rs1861867&quot;, etc.">
  </div>
  <div class="col-xs-12 col-sm-4 col-sm-6">
    <h5 class="pull-right"><div id="streamtable-found" class="label label-primary"></div></h5>
  </div>
</div>
<div class="row stream_table_row">
  <div class="col-xs-12">
    <table id="stream_table" class="table table-striped table-bordered">
      <thead>
        <tr>
          <th>Variant</th>
          <th>Nearest Gene(s)</th>
          <th>MAF</th>
          <th>P-value</th>
          <th>Effect Size (se)</th>
        </tr>
      </thead>
      <tbody>
      </tbody>
    </table>
  </div>
</div>

<div class="row">
  <h3 id="qq">QQ plot:</h3>
    <div class="pheno-info col-xs-12 col-sm-4 col-md-3 col-lg-3">
        <div id="qq_plot_container_both">All</div>
        <p class="gc-control_both"></p>
    </div>
    {% if sex_stratified %}
      {% if pheno_male != None %}
        <div class="pheno-info col-xs-12 col-sm-4 col-md-3 col-lg-3">
          <div id="qq_plot_container_male">Male</div>
          <p class="gc-control_male"></p>
        </div>
      {% endif %}
      {% if pheno_female != None %}
        <div class="pheno-info col-xs-12 col-sm-4 col-md-3 col-lg-3">
          <div id="qq_plot_container_female">Female</div>
          <p class="gc-control_female"></p>
        </div>
      {% endif %}
    {% endif %}
</div>
<i>(Genomic Control lambda calculated based on the 50th percentile (median), 10th percentile, 1st percentile, and 1/10th of a percentile)</i>


{% if show_correlations %}
  <div class="row" id="correlations">
    <div class="col-md-12">
      <h2>Correlated phenotypes for {% include "pheno/h1.html" %}</h2>
      <p>
        Genetically correlated traits are calculated using genome-wide association summary statistics via LD Score regression (<a href="https://doi.org/10.1038/ng.3211">LDSC</a>).
      </p>
      <div id="correlations-table" class="table-striped table-bordered"></div>
    </div>
  </div>
{% endif %}


<script type="text/template" id="streamtable-template">
<tr>
<td><a style="color:black" href="{{ url_for('.variant_page', query='').rstrip('/') }}/<%= v.chrom %>-<%= v.pos %>-<%= v.ref %>-<%= v.alt %>">
    <%= v.chrom %>:<%= v.pos.toLocaleString() %> <%= v.ref %> / <%= v.alt %>
    <% if (v.rsids) { %>(<%= v.rsids.replace(/,/g, ', ') %>)<% } %>
</a></td>
<td><i><%= v.nearest_genes.replace(/,/g, ', ') %></i></td>
<td><%= (v.maf)? v.maf.toPrecision(2) : (v.af)? Math.min(v.af, 1-v.af).toPrecision(2) : (v.ac)? v.ac : "" %></td>
<td><%= (v.pval==0) ? '≤1e-320' : v.pval.toExponential(1) %></td>
<td><%= (v.beta) ? v.beta.toPrecision(2) : "" %><%= (v.sebeta) ? " ("+v.sebeta.toPrecision(2)+")" : "" %></td>
</tr>
</script>
{% endblock %}
