{% extends "layout.html" %}

{% block in_head %}
<script src="https://cdn.jsdelivr.net/npm/d3@^5.16.0" type="text/javascript"></script>
<script src="https://unpkg.com/d3-tip@0.9.1"></script>
<script src="{{ url_for('.static', filename='vendor/stream_table-1.1.1.min.js') }}" type="text/javascript"></script>
<script src="{{ url_for('.static', filename='pheno.js') }}" type="text/javascript"></script>
{% if show_correlations %}
  {#  TODO: Better dependency management #}
  <link href="https://unpkg.com/tabulator-tables@4.2.2/dist/css/bootstrap/tabulator_bootstrap.min.css" rel="stylesheet">
  <script type="text/javascript" src="https://unpkg.com/tabulator-tables@4.2.2/dist/js/tabulator.min.js"></script>
{% endif %}

<script type="text/javascript">
  window.model.tooltip_underscoretemplate = {{ tooltip_underscoretemplate|tojson }};
  window.pheno = {{ phenocode|tojson|safe }};
  window.model.sex_stratified = {{ sex_stratified | tojson }};
  window.model.show_correlations = {{ show_correlations | tojson }};
  window.model.correlations_url = {{ url_for('.api_pheno_correlations', phenocode=pheno['phenocode'] ) | tojson }};
  window.model.pheno_correlations_pvalue_threshold = {{ pheno_correlations_pvalue_threshold | tojson }};

  //get max qq plot height and width value from combined, male and female jsons.
  window.model.qq_plot_height = 0;
  window.model.qq_plot_width = 0;  
  var deferredObjects = [];

  if (window.model.sex_stratified){
    // Add JSON calls
    addJsonCall(window.model.urlprefix + "/api/qq/pheno/" + window.pheno + ".json");
    addJsonCall(window.model.urlprefix + "/api/qq/pheno/" + window.pheno_male + ".json");
    addJsonCall(window.model.urlprefix + "/api/qq/pheno/" + window.pheno_female + ".json");

    // Wait for all JSON calls to finish
    $.when.apply($, deferredObjects).done(function(result1, result2, result3) {
      // result1, result2, result3 correspond to the data from file1.json, file2.json, and file3.json respectively
      var data1 = result1[0];
      var data2 = result2[0];
      var data3 = result3[0];

      // Access and work with data1, data2, data3 here
      console.log('Data from file1.json:', data1);
      console.log('Data from file2.json:', data2);
      console.log('Data from file3.json:', data3);

      // Proceed with your code here
  });
  } else {
    $.getJSON(window.model.urlprefix + "/api/qq/pheno/" + window.pheno + ".json")
    .done(function(data) {
      window.debug.qq = data;
      console.log(window.debug.qq)
      window.debug.qq.by_maf.forEach(function(maf_ranges){
        console.log(maf_ranges.qq.bins)
        max_height = maf_ranges.qq.bins[maf_ranges.qq.bins.length - 1][1]
        max_width = maf_ranges.qq.bins[maf_ranges.qq.bins.length - 1][0]

        if (window.model.qq_plot_height < max_height ){
          window.model.qq_plot_height = max_height
        }
        if (window.model.qq_plot_width < max_width){
          window.model.qq_plot_width = max_width
        }
      });
    });
  }

  if (window.model.sex_stratified){
    $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.pheno + ".male.json")
    .done(function(data) {
      window.debug.qq = data;
      console.log(window.debug.qq)
      window.debug.qq.by_maf.forEach(function(maf_ranges){
        console.log(maf_ranges.qq.bins)
        max_height = maf_ranges.qq.bins[maf_ranges.qq.bins.length - 1][1]
        max_width = maf_ranges.qq.bins[maf_ranges.qq.bins.length - 1][0]

        if (window.model.qq_plot_height < max_height ){
          window.model.qq_plot_height = max_height
        }
        if (window.model.qq_plot_width < max_width){
          window.model.qq_plot_width = max_width
        }
      });
    });
    $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.pheno + ".female.json")
    .done(function(data) {
      window.debug.qq = data;
      console.log(window.debug.qq)
      window.debug.qq.by_maf.forEach(function(maf_ranges){
        console.log(maf_ranges.qq.bins)
        max_height = maf_ranges.qq.bins[maf_ranges.qq.bins.length - 1][1]
        max_width = maf_ranges.qq.bins[maf_ranges.qq.bins.length - 1][0]

        if (window.model.qq_plot_height < max_height ){
          window.model.qq_plot_height = max_height
        }
        if (window.model.qq_plot_width < max_width){
          window.model.qq_plot_width = max_width
        }
      });
    });
  }
  

  $.getJSON(window.model.urlprefix + "/api/manhattan/pheno/" + window.pheno + ".json")
  .done(function(data) {
      window.debug.manhattan = data;
      create_gwas_plot(data.variant_bins, data.unbinned_variants);
      populate_streamtable(data.unbinned_variants);
  })
  .fail(function() {
    console.log("Manhattan XHR failed");
  });

  if (window.model.sex_stratified){
    $.when(
      $.getJSON(window.model.urlprefix + "/api/miami/pheno/" + window.pheno_male + ".json"),
      $.getJSON(window.model.urlprefix + "/api/miami/pheno/" + window.pheno_female + ".json")
    ).done(function(data_male, data_female) {
        window.debug.miami = {
          male: data_male[0],
          female: data_female[0]
        };
        create_miami_plot(
          data_male[0].variant_bins, data_male[0].unbinned_variants,
          data_female[0].variant_bins, data_female[0].unbinned_variants
        );
        // we won't worry about this for now.
        // populate_streamtable(data_male.unbinned_variants);
    })
    .fail(function() {
      console.log("Miami XHR failed");
    });
  }
  
  $.getJSON(window.model.urlprefix + "/api/qq/pheno/" + window.pheno + ".json")
  .done(function(data) {
    window.debug.qq = data;
    _.sortBy(_.pairs(data.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
      var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
      if (i === 0) { text = '<b>' + text + '</b>'; }
      text = '<br>' + text;
      $('.gc-control_both').append(text);
    });
  if (data.by_maf)
    create_qq_plot(data.by_maf, data.ci, window.model.qq_plot_height, window.model.qq_plot_width);
  else
    create_qq_plot([{maf_range:[0,0.5], qq:data.overall.qq, count:data.overall.count}], data.ci, window.model.qq_plot_height, window.model.qq_plot_width);
  })
  .fail(function() {
    console.log("QQ XHR failed");
  });

  if (window.model.sex_stratified){
      //male
      $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.pheno + ".male.json")
      .done(function(data) {
      window.debug.qq_male = data;
      _.sortBy(_.pairs(data.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
        var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
        if (i === 0) { text = '<b>' + text + '</b>'; }
        text = '<br>' + text;
        $('.gc-control_male').append(text);
      });
    if (data.by_maf)
      create_qq_plot(data.by_maf, data.ci, window.model.qq_plot_height, window.model.qq_plot_width, "male");
    else
      create_qq_plot([{maf_range:[0,0.5], qq:data.overall.qq, count:data.overall.count}], data.ci, window.model.qq_plot_height, window.model.qq_plot_width, "male");
    })
    .fail(function() {
      console.log("QQ male XHR failed");
    });
    //female
    $.getJSON(window.model.urlprefix + "/api/qq/miami/pheno/" + window.pheno + ".female.json")
      .done(function(data) {
      window.debug.qq_male = data;
      _.sortBy(_.pairs(data.overall.gc_lambda), function(d) {return -d[0];}).forEach(function(d, i) {
        var text = 'GC lambda ' + d[0] + ': ' + d[1].toFixed(3);
        if (i === 0) { text = '<b>' + text + '</b>'; }
        text = '<br>' + text;
        $('.gc-control_female').append(text);
      });
    if (data.by_maf)
      create_qq_plot(data.by_maf, data.ci, window.model.qq_plot_height, window.model.qq_plot_width, "female");
    else
      create_qq_plot([{maf_range:[0,0.5], qq:data.overall.qq, count:data.overall.count}], data.ci, window.model.qq_plot_height, window.model.qq_plot_width, "female");
    })
    .fail(function() {
      console.log("QQ female XHR failed");
    });
  }
</script>
<style type="text/css">
.d3-tip {
  line-height: 1.4;
  padding: 12px;
  background: rgba(0, 0, 0, 0.8);
  color: #fff;
  border-radius: 4px;
  pointer-events: none;
}
/* Creates a small triangle extender for the tooltip */
.d3-tip:after {
  display: inline;
  font-size: 10px;
  width: 100%;
  line-height: 1;
  color: rgba(0, 0, 0, 0.8);
  position: absolute;
  pointer-events: none;
}
/* Northward tooltips */
.d3-tip.n:after {
  content: "\25BC";
  margin: -3px 0 0 0;
  top: 100%;
  left: 0;
  text-align: center;
}

#manhattan_plot_container {
  min-width: 700px;
}
#manhattan_plot_container .axis > path.domain {
  stroke-width: 2px;
  stroke: #666;
  fill: none;
}
#manhattan_plot_container .axis g.tick line {
  stroke: #666;
}

#qq_plot_container_both .axis path.domain {
  stroke: black;
  fill: none;
  text-align: center;
}
#qq_plot_container_both g.tick line {
  stroke: #666;
  opacity: 0.3;
}

#qq_plot_container_male .axis path.domain {
  stroke: black;
  fill: none;
  text-align: center;
}
#qq_plot_container_male g.tick line {
  stroke: #666;
  opacity: 0.3;
}

#qq_plot_container_female .axis path.domain {
  stroke: black;
  fill: none;
  text-align: center;
}
#qq_plot_container_female g.tick line {
  stroke: #666;
  opacity: 0.3;
}

.pheno-info > p {
  margin-bottom: 0;
}

.section-link {
  margin-right: 20px;
  display: inline;
}

.tabulator .tabulator-footer .tabulator-page.active {
  background-color: #337ab7;
  color: white;
}

.parent {
  background: mediumpurple;
  padding: 1rem
}

.child {
  border: 1px solid indigo;
  padding: 1rem
}

.inline-block-child {
  display: inline-block;
}

</style>
{% endblock %}


{% block contained %}
{% for key in ['num_cases', 'num_controls', 'num_samples'] %}
 {% if key in pheno %}
  {% if '<' in pheno[key]|string %}
    <div class="alert alert-danger" role="alert"><b>Warning:</b> This phenotype has {{ pheno[key] }} {{ key.replace('num_','') }}.</div>
  {% elif 0 < pheno[key]|int < 200 %}
    <div class="alert alert-danger" role="alert"><b>Warning:</b> This phenotype only has {{ pheno[key] }} {{ key.replace('num_', '') }}.</div>
  {% endif %}
 {% endif %}
{% endfor %}
<br>
<div class="row">
    <div class="col-xs-12">
        <h1 style="margin-top:0">{% include "pheno/h1.html" %}</h1>
    </div>
</div>
<div class="row">
  <div class="pheno-info col-xs-12 col-sm-8">
      {% if 'num_cases' in pheno and pheno['num_cases'] != "" %}
        <p><b>{{ pheno.num_cases }}</b> cases, <b>{{ pheno.num_controls }}</b> controls</p>
      {% elif 'num_samples' in pheno %}
        <p><b>{{ pheno.num_samples }}</b> samples</p>
      {% endif %}
      {% if pheno.category %}
        <p>Category: <b>{{ pheno.category}}</b></p>
      {% endif %}
      {% include "pheno/info.html" %}
  </div>
  <div class="pheno-info col-xs-12 col-sm-4">
    {% if config.DOWNLOAD_PHENO_SUMSTATS_BUTTON %}
    <span class="pull-right">
      <a class="btn btn-default" href="{{ url_for('.download_pheno', phenocode=phenocode) }}">Download summary statistics</a>
    </span>
    {% endif %}
    {% if show_manhattan_filter_button %}
    <span class="pull-right" style="margin-left:1em;margin-right:1em">
      <a class="btn btn-default" href="{{ url_for('.pheno_filter_page', phenocode=phenocode) }}">Filter Variants</a>
    </span>
    {% endif %}
  </div>
</div>

<div class="row">
    <div class="col-xs-12" id="manhattan">
        <div id="manhattan_plot_container">
        </div>
    </div>
</div>

<div class="row">
  <div class="col-xs-12 col-md-8">
    <h3>Top Loci:</h3>
  </div>
</div>
<div class="row">
  <div class="col-xs-12 col-sm-8 col-md-6">
    <input type="text" class="form-control" id="search" placeholder="Search... &quot;TCF7L2&quot;, &quot;rs1861867&quot;, etc.">
  </div>
  <div class="col-xs-12 col-sm-4 col-sm-6">
    <h5 class="pull-right"><div id="streamtable-found" class="label label-primary"></div></h5>
  </div>
</div>

<div class="row stream_table_row">
  <div class="col-xs-12">
    <table id="stream_table" class="table table-striped table-bordered">
      <thead>
        <tr>
          <th>Variant</th>
          <th>Nearest Gene(s)</th>
          <th>MAF</th>
          <th>P-value</th>
          <th>Effect Size (se)</th>
        </tr>
      </thead>
      <tbody>
      </tbody>
    </table>
  </div>
</div>

<div class="row">
  <h3 id="qq">QQ plot:</h3>
    <div class="pheno-info col-xs-12 col-sm-4 col-md-3 col-lg-3">
        <div id="qq_plot_container_both">All</div>
        <p class="gc-control_both"></p>
    </div>
    {% if sex_stratified %}
    <div class="pheno-info col-xs-12 col-sm-4 col-md-3 col-lg-3">
      <div id="qq_plot_container_male">Male</div>
      <p class="gc-control_male"></p>
    </div>
    <div class="pheno-info col-xs-12 col-sm-4 col-md-3 col-lg-3">
      <div id="qq_plot_container_female">Female</div>
      <p class="gc-control_female"></p>
    </div>
    {% endif %}
</div>
<i>(Genomic Control lambda calculated based on the 50th percentile (median), 10th percentile, 1st percentile, and 1/10th of a percentile)</i>


{% if show_correlations %}
  <div class="row" id="correlations">
    <div class="col-md-12">
      <h2>Correlated phenotypes for {% include "pheno/h1.html" %}</h2>
      <p>
        Genetically correlated traits are calculated using genome-wide association summary statistics via LD Score regression (<a href="https://doi.org/10.1038/ng.3211">LDSC</a>).
      </p>
      <div id="correlations-table" class="table-striped table-bordered"></div>
    </div>
  </div>
{% endif %}


<script type="text/template" id="streamtable-template">
<tr>
<td><a style="color:black" href="{{ url_for('.variant_page', query='').rstrip('/') }}/<%= v.chrom %>-<%= v.pos %>-<%= v.ref %>-<%= v.alt %>">
    <%= v.chrom %>:<%= v.pos.toLocaleString() %> <%= v.ref %> / <%= v.alt %>
    <% if (v.rsids) { %>(<%= v.rsids.replace(/,/g, ', ') %>)<% } %>
</a></td>
<td><i><%= v.nearest_genes.replace(/,/g, ', ') %></i></td>
<td><%= (v.maf)? v.maf.toPrecision(2) : (v.af)? Math.min(v.af, 1-v.af).toPrecision(2) : (v.ac)? v.ac : "" %></td>
<td><%= (v.pval==0) ? '≤1e-320' : v.pval.toExponential(1) %></td>
<td><%= (v.beta) ? v.beta.toPrecision(2) : "" %><%= (v.sebeta) ? " ("+v.sebeta.toPrecision(2)+")" : "" %></td>
</tr>
</script>
{% endblock %}
