<div class="modal fade" id="imagemodal" tabindex="-1" style="width: 100%;" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
  <div class="modal-dialog" style="width: 60%;">
    <div class="modal-content" style="width: 100%;">              
      <div class="modal-body" style="width: 100%;">
        <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
        <img src="" class="imagepreview" style="width: 100%;" >
      </div>
    </div>
  </div>
</div>
<!-- Menu -->
<div class="col-xs-3 col-sm-3 col-md-3 col-lg-3" role='complementary'>
    <ul id="navigator" class="nav nav-tabs nav-stacked text-left" aria-orientation="vertical"  role="tablist">
        <li class="active"><a data-toggle="tab" href="#Overview"><b>Overview</b></a></li>
        <li><a data-toggle="tab" href="#GenotypeData"><b>1. Genotype Data</b></a></li>
        <li><a data-toggle="tab" href="#LiftOver"><b>2. LiftOver</b></a></li>
        <li><a data-toggle="tab" href="#PreLiftOver">2.1. Pre-LiftOver Filtering</a></li>
        <li><a data-toggle="tab" href="#LiftOverGRCh38">2.2. LiftOver to GRCh38</a></li>
        <li><a data-toggle="tab" href="#HweRuth">2.3. Ancestry-Aware Hardy-Weinberg Check with RUTH</a></li>
        <li><a data-toggle="tab" href="#AncestryEuro"><b>3. Ancestry Check and European Subset</b></a></li>
        <li><a data-toggle="tab" href="#CommonVariantFiltering"><b>4. Common Variant Filtering</b></a></li>
        <li><a data-toggle="tab" href="#TOPMedImputation"><b>5. TOPMed Imputation</b></a></li>
        <li><a data-toggle="tab" href="#PhenotypeAssociation"><b>6. Phenotype Association</b></a></li>
        <li><a data-toggle="tab" href="#BinaryPhenotype">6.1 Selecting Binary Phenotypes</a></li>
        <li><a data-toggle="tab" href="#ContinuousPhenotype">6.2 ContinuousPhenotype</a></li>
        <li><a data-toggle="tab" href="#Regenie">6.3 Regenie</a></li>
        <li><a data-toggle="tab" href="#PheWeb"><b>7. PheWeb</b></a></li>
        <li><a data-toggle="tab" href="#Preperation">7.1 Preperation</a></li>
        <li><a data-toggle="tab" href="#SetUpHosting">7.2 Set Up and Hosting</a></li>
        <li><a data-toggle="tab" href="#Acknowledgements"><b>Acknowledgements</b></a></li>
    </ul>
</div>
<div class="col-xs-9 col-sm-9 col-md-9 col-lg-9">
    <div class="tab-content">
<!-- OVERVIEW SECTION-->
        <div id="Overview" class="tab-pane fade in active">
            <h2 class="text-center"> About CLSA PheWeb</h2>
            <hr></hr>
            <p> This platform presents multiple analyses conducted using CLSA data to the scientific community as well as the public at large. The <a href="https://www.clsa-elcv.ca/">CLSA</a> cohort is specific to Canada. </p>
            <p> This platform displays Phenome-wide association (PheWAS) results from the CLSA cohort (i.e. statistical associations between individuals' trait information and genomic variations across the cohort).  </p>
            <p style='margin-right:0cm;margin-left:0cm;font-size:16px;font-family:"Calibri",sans-serif;margin:0cm;text-align:center;'><span style='font-family:"Times New Roman",serif;'>&nbsp;</span></p>
            <img src="static/images/CLSA-logo-1.png" width="60%" height="60%" class="center-block">
            <p style='margin-right:0cm;margin-left:0cm;font-size:16px;font-family:"Calibri",sans-serif;margin:0cm;text-align:center;'><span style='font-family:"Times New Roman",serif;'>&nbsp;</span></p>
        </div>
<!-- DATA DESCRIPTION -->
    <div id="GenotypeData" class="tab-pane fade">
        <h2 class="text-center">Genotype Data</h2>
        <hr></hr>
        <p>Original non-imputed CLSA genotype data has 794,409 variants for 26,626 participants.</p>
        <p>Participants were genotyped using the Affrymetrix UK Biobank Axiom array.</p>
        <p>Originally, all genomic positions are reported in reference to human genome build GRCh37.</p>
        <p>Participants are 51% female between the ages of 45 and 85 years when recruited.</p>
    </div>

<!-- LiftOver and CLSA filtering Overview-->
        <div id="LiftOver" class="tab-pane fade">
            <h2 class="text-center">Pre-imputation</h2>
            <p>liftOver scripts and in-depth description of usage available on <a href="https://github.com/GaglianoTaliun-Lab/CLSA_PheWeb/tree/main/GRCh37_to_GRCh38" target="_blank">Github</a></p>
            <hr></hr>
            <div id="PreLiftOver">
                <p>Before liftOver, we used the CLSA's quality checks for marker and sample quality control.</p>
                <p>The CLSA filtered out variants with:</p>
                <ol>
                    <li>Batch Discordance</li>
                    <li>Hardy-Weinberg Discordance (p < 3.15e-10)</li>
                    <li>Control Genotype Discordance </li>
                    <li>Sex Discordance</li>
                    <li>Low Minor Allele Frequency (MAF < 0.05)</li>
                    <li>Indels</li>
                </ol>
                <p>This removed 131,892 variants, keeping 662,517.</p>
                <hr></hr>
                <p>We removed samples with:</p>
                <ol>
                    <li>Identification as an outlier in relation to heterozygosity or genotype missingness</li>
                    <li>Differences between reported sex and chromosomal sex</li>
                    <li>Missing self-reported sex </li>
                    <li>Missing chromosomal sex</li>
                </ol>
                <p>This removed 66 samples, keeping 26,563</p>
            </div>
            <hr></hr>
            <div id="LiftOverGRCh38">
                <p>We converted genotype array data from genome build GRCh37 to GRCh38, aligned to the "+" strand using LiftOver. The conversion included the following steps:</p>
                <ol>
                    <li>Batch Discordance</li>
                    <li>Hardy-Weinberg Discordance (p < 3.15e-10)</li>
                    <li>Control Genotype Discordance </li>
                    <li>Sex Discordance</li>
                    <li>Low Minor Allele Frequency (MAF < 0.05)</li>
                    <li>Indels</li>
                </ol>
                <p>This removed 131,892 variants, keeping 662,517.</p>
                <hr></hr>
                <p>We removed samples with:</p>
                <ol>
                    <li>Identification as an outlier in relation to heterozygosity or genotype missingness</li>
                    <li>Differences between reported sex and chromosomal sex</li>
                    <li>Missing self-reported sex </li>
                    <li>Missing chromosomal sex</li>
                </ol>
                <p>This removed 66 samples, keeping 26,563</p>
            </div>
            <div id="HweRuth">
                <p>Using Robust Unified Hardy-Weingberg Equilibrium Test (<a href="https://github.com/statgen/ruth" target = "_blank" >RUTH</a>), we confirmed the CLSA's hardy-weinberg equilibrium exclusion criteria.</p>
                <p>We ran RUTH on the results from step 2.2 using the 20 first PCs using the likelihood-ratio test with max lambda parameter of 0, using the GT field. We used seed 101.</p>
                <p>No additional variants were excluded using this test.</p>
            </div>
        </div>

<!-- Ancestry Check and European Subset  -->
        <div id="AncestryEuro" class="tab-pane fade">
            <h2 class="text-center">Lift over to GRCh38</h2>
            <hr></hr>
            <p>We ran our own ancestry check and filtering using a custom pipeline. The pipeline can be found <a href="https://github.com/CERC-Genomic-Medicine/HGDP_1KG_ancestry_inference" target="_blank">here</a></p>
            <p>We subset individuals of predicted european ancestry. For ancestry prediction we followed the following steps:</p>
            <ol>
                <li>We used <a href="https://csg.sph.umich.edu/chaolong/LASER/TRACE_Manual.pdf" target = "_blank">TRACE</a> to project principal components onto HGDP and 1KG combined references (Figure 1). We removed individuals that had a Z score of <span>&#177;</span>5.</li>
                <li>Using the projection results, we used a RandomForestModel to calculate probabilities of ancestry group.</li>
                <li>We decided to use 0.78 probability as our threshold to include in european subset.</li>
            </ol>
            <p>In the end, this resultied in 24505 individuals being included with the following distribution of variants per chromosome:</p>
            <table class="table table-bordered table-condensed">
                <tr>
                    <th>Chromosome</th>
                    <th>Number of SNPs</th>
                </tr>
                <tr>
                    <td>1</td>
                    <td>47826</td>
                </tr>
                <tr>
                    <td>2</td>
                    <td>48686</td>
                </tr>
                <tr>
                    <td>3</td>
                    <td>40648</td>
                </tr>
                <tr>
                    <td>4</td>
                    <td>38662</td>
                </tr>
                <tr>
                    <td>5</td>
                    <td>36654</td>
                </tr>
                <tr>
                    <td>6</td>
                    <td>42563</td>
                </tr>
                <tr>
                    <td>7</td>
                    <td>33200</td>
                </tr>
                <tr>
                    <td>8</td>
                    <td>31200</td>
                </tr>
                <tr>
                    <td>9</td>
                    <td>26652</td>
                </tr>
                <tr>
                    <td>10</td>
                    <td>30518</td>
                </tr>
                <tr>
                    <td>11</td>
                    <td>29892</td>
                </tr>
                <tr>
                    <td>12</td>
                    <td>28890</td>
                </tr>
                <tr>
                    <td>13</td>
                    <td>20865</td>
                </tr>
                <tr>
                    <td>14</td>
                    <td>19620</td>
                </tr>
                <tr>
                    <td>15</td>
                    <td>18916</td>
                </tr>
                <tr>
                    <td>16</td>
                    <td>21012</td>
                </tr>
                <tr>
                    <td>17</td>
                    <td>19726</td>
                </tr>
                <tr>
                    <td>18</td>
                    <td>18143</td>
                </tr>
                <tr>
                    <td>19</td>
                    <td>16562</td>
                </tr>
                <tr>
                    <td>20</td>
                    <td>15848</td>
                </tr>
                <tr>
                    <td>21</td>
                    <td>9102</td>
                </tr>
                <tr>
                    <td>22</td>
                    <td>9661</td>
                </tr>
                <tr>
                    <td>23</td>
                    <td>15503</td>
                </tr>
            </table>
        </div>

<!-- Common Variant Filtering -->
        <div id="CommonVariantFiltering" class="tab-pane fade">
            <h2 class="text-center">Common Variant Filtering</h2>
            <hr></hr>
            <p>After filtering european sub-population, we proceeded to further filter common non-imputed variants for step 1 of regenie.</p>
            <p>We performed the following:</p>
            <ol>
                <li>Removed variants with sample missingness of > 0.1</li>
                <li>Removed samples with genotype missingness of > 0.1</li>
                <li>Variants departing from Hardy-Weinberg equilibrium with a p-value less than 1e-15</li>
                <li>Linkage disequilibrium pruning of correlated variants with window size of 1000, step size of 100 and $r^2$ threshold of 0.9</li>
            </ol>
            <p>At the end, we had this distribution of variants per chromosome for 24505 individuals:</p>
            <table class="table table-bordered table-condensed">
                <tr>
                    <th>Chromosome</th>
                    <th>Number of SNPs</th>
                </tr>
                <tr>
                    <td>1</td>
                    <td>26319</td>
                </tr>
                <tr>
                    <td>2</td>
                    <td>26537</td>
                </tr>
                <tr>
                    <td>3</td>
                    <td>21961</td>
                </tr>
                <tr>
                    <td>4</td>
                    <td>20823</td>
                </tr>
                <tr>
                    <td>5</td>
                    <td>20043</td>
                </tr>
                <tr>
                    <td>6</td>
                    <td>19441</td>
                </tr>
                <tr>
                    <td>7</td>
                    <td>18494</td>
                </tr>
                <tr>
                    <td>8</td>
                    <td>16440</td>
                </tr>
                <tr>
                    <td>9</td>
                    <td>15021</td>
                </tr>
                <tr>
                    <td>10</td>
                    <td>17505</td>
                </tr>
                <tr>
                    <td>11</td>
                    <td>16159</td>
                </tr>
                <tr>
                    <td>12</td>
                    <td>15601</td>
                </tr>
                <tr>
                    <td>13</td>
                    <td>11919</td>
                </tr>
                <tr>
                    <td>14</td>
                    <td>11202</td>
                </tr>
                <tr>
                    <td>15</td>
                    <td>10768</td>
                </tr>
                <tr>
                    <td>16</td>
                    <td>11953</td>
                </tr>
                <tr>
                    <td>17</td>
                    <td>11463</td>
                </tr>
                <tr>
                    <td>18</td>
                    <td>10701</td>
                </tr>
                <tr>
                    <td>19</td>
                    <td>8816</td>
                </tr>
                <tr>
                    <td>20</td>
                    <td>8675</td>
                </tr>
                <tr>
                    <td>21</td>
                    <td>5338</td>
                </tr>
                <tr>
                    <td>22</td>
                    <td>5667</td>
                </tr>
                <tr>
                    <td>23</td>
                    <td>7860</td>
                </tr>
            </table>
            <p>For a total of 338706 SNPs.</p>
        </div>


<!-- TOPMed Imputation -->
        <div id="TOPMedImputatione" class="tab-pane fade">
            <h2 class="text-center">TOPMed Imputation</h2>
            <hr></hr>
        </div>

<!-- Phenotype Association -->
        <div id="PhenotypeAssociation" class="tab-pane fade">
            <h2 class="text-center">Phenotype Association</h2>
            <hr></hr>
            <div id="BinaryPhenotype">

            </div>
            <div id="ContinuousPhenotype">
                <p>Continous phenotypes were selected with the following criteria:</p>
                <ol>
                    <li>More than 20 uniques numeric values</li>
                    <li>More than 1000 total numeric values</li>
                    <li>One value accounted for more than 10% of all values</li>
                    <li>Removed outliers $\pm$ 5 standard deviations from the mean</li>
                    <li>Removed traits inappropriate for a genetic association analysis</li>
                </ol>
                <p>The final number of continuous phenotypes was 357.</p>
            </div>
            <div id="Regenie">
                <p>Association testing was performed using <a href="https://rgcgithub.github.io/regenie/" target="_blank">Regenie</a>.</p>
                <p>Variants from step 4 were used in step 1. Variants from step 5 were used in step 2.</p>
                <p>For binary variables, options ____ were used for step 1, and ___ for step 2.</p>
                <p>For continuous variables, rank inverse normal transformation was used for both steps with bsize of 1000.</p>
            </div>
        </div>

    <!-- PheWeb  -->
        <div id="PheWeb" class="tab-pane fade">
            <h2 class="text-center">PheWeb</h2>
            <hr></hr>
            <div id="Preperation"></div>
            <div id="SetUpHosting"></div>
        </div>
     <!-- Acknowledgements  -->
        <div id="Acknowledgements" class="tab-pane fade">
            <h2 class="text-center">Acknowledgements</h2>
           <hr></hr>
            <ul>
                <li>Authors : </li>
                <li>All computational analyses were performed using compute resources of Digital Research Alliance of Canada, and the Digital Research Alliance of Canada hosts this site.</li>
                <li>We thank all CLSA participants who volunteered their time and information for research.</li>
            </ul>

        </div>
    <!-- end of file -->
    </div>
</div>
